src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,168 @@
+   :ArrowUp     [:position (:up    -motions)]
+   :ArrowLeft   [:position (:left  -motions)]
+   :ArrowRight  [:position (:right -motions)]
+
+   :KeyW        [:position [ 4  0  0]]
+   :KeyB        [:position [-4  0  0]]
+
+   :ShiftBracketRight [:frame  1]
+   :BracketRight      [:frame -1]
+
+   ;; absolute movement
+   :KeyG        [:position [:first :first :first]]
+   :ShiftKeyG   [:position [:last  :last  :last]]
+   :Digit0      [:position [:first :first :first]]
+   :ShiftDigit4 [:position [:last  :last  :last]]
+
+   ;; mode change
+   :KeyI [:mode :edit]
+   :KeyV [:mode :visual]
+
+   ;; other
+   :KeyX        [:macro {:type :simple :value [[:attr nil] (:down-line -motions)]}]
+   ;; so these could be special or just, like... :tempo, and :octave
+   ;; I'm leaving them out for now
+   ;; :Period [:special [:tempo 1]]
+   ;; :ShiftPeriod [:special [:tempo -1]]
+   :Space [:special :play-pause]})
+
+(def edit-keymap
+  ;; hm, this isn't going to work quite like you want it to.
+  ;; this is fine for inputting notes...
+  {:KeyA [:attr :C]    :KeyW [:attr :C#]
+   :KeyS [:attr :D]    :KeyE [:attr :D#]
+   :KeyD [:attr :E]    :KeyR [:attr :F]
+   :KeyF [:attr :F#]   :KeyT [:attr :G]
+   :KeyG [:attr :G#]   :KeyY [:attr :A]
+   :KeyH [:attr :A#]   :KeyU [:attr :B]
+
+   ;; ... but when you get to the other attributes, what happens? They have
+   ;; values in [0x00, 0xff].
+   ;; You gotta either have another sort of translation layer where some of your
+   ;; note names become hex digits, or... something
+   ;; In fact, now that I think about it, this is why volume in the pre-alpha version
+   ;; only goes up to 9.
+   :Digit0 [:attr :Digit0]
+   :Digit1 [:attr :Digit1]
+   :Digit2 [:attr :Digit2]
+   :Digit3 [:attr :Digit3]
+   :Digit4 [:attr :Digit4]
+   :Digit5 [:attr :Digit5]
+   :Digit6 [:attr :Digit6]
+   :Digit7 [:attr :Digit7]
+   :Digit8 [:attr :Digit8]
+   :Digit9 [:attr :Digit9]
+
+   :KeyX      :off
+   :ShiftKeyX :stop
+
+   :Backspace [:macro {:type :simple :value [[:attr nil] (:up-line -motions)]}]
+   :Space     [:position (:down-line -motions)]})
+
+(def keymap
+  {:normal normal-keymap
+   :edit   edit-keymap})
+
+(def -movement-keys
+  #{:Space :ArrowDown :ArrowUp :ArrowLeft :ArrowRight :Tab :Backspace})
+
+(defn prevent-movement! [ev]
+  "Prevent movement keys (arrows, space, tab) from moving focus or scrolling."
+  (when (-movement-keys (keyword (.-code ev)))
+    (.preventDefault ev))
+  ev)
+
+(defn translate-keycode [ev]
+  "Takes keydown event and returns keycode as keyword; e.g., :KeyI, :Digit0.
+  If SHIFT is held, prepend Shift; e.g., :ShiftKeyI.
+  If CTRL is held, prepend Ctrl; e.g., :CtrlShiftKeyI."
+  (keyword (str (when (.-ctrlKey ev) "Ctrl")
+                (when (.-shiftKey ev) "Shift")
+                ;; .-code is nil if not a keypress
+                (.-code ev))))
+
+;; XXX don't know what to name these, but might not matter
+(defn position!
+  [params state])
```

Comment:
probably want these to return seqs so that it's easy to support actions with multiple consequences e.g. macros, visual mode operations, etc.

Reply:
Now that these are more filled out, I have general purpose state-changing code, like this
https://github.com/cqsd/chipper/blob/86ec97dda246eeffeffbe86c90cad5bf6d5752e0/src/chipper/keyboard2.cljs#L145-L152 . One will exist for every primitive operation on the state, and be maximally explicit (e.g. `set-attr!` will be roughly like `(set-attr! [attr position state] ... )`). Usable for more complex operations.